File: lc/coreprep/implementation/practice_iterations/two_pointers_4/SubarraysWithProductLessThanTarget.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n3) / O(1)
public class SubarraysWithProductLessThanTarget {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int s = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        List<List<Integer>> res = new ArrayList<>();
        int windowStart = 0;
        int product = 1;

        for (int windowEnd = 0; windowEnd < n; windowEnd++) {
          product *= a[windowEnd];

          while (product >= s) {
            product /= a[windowStart];
            windowStart++;
          }

          List<Integer> lst = new LinkedList<>();
          for (int i = windowEnd; i >= windowStart; i--) {
            lst.add(0, a[i]);
            res.add(new ArrayList<>(lst));
          }
        }

        for (var sub : res) {
          for (int e : sub) {
            System.out.printf("%d ", e);
          }
          System.out.println();
        }
      }
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/two_pointers_3/subarrays_with_product_less_than_target.cpp
#include <algorithm>
#include <iostream>
#include <list>
#include <vector>

using namespace std;

// O(n3) / O(n2)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, s;
  cin >> tt;

  while (tt--) {
    cin >> n >> s;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    vector<vector<int>> res;
    int product = 1;
    int window_start = 0;

    for (int window_end = 0; window_end < n; window_end++) {
      product *= a[window_end];

      while (product >= s) {
        product /= a[window_start];
        window_start++;
      }

      list<int> lst;
      for (int i = window_end; i >= window_start; i--) {
        lst.push_front(a[i]);

        vector<int> v(lst.begin(), lst.end());
        res.push_back(v);
      }
    }

    for (auto sub : res) {
      for (int e : sub) {
        cout << e << " ";
      }
      cout << "\n";
    }
  }

  return 0;
}